appendix-basic65.tex

```diff
@@ -220,168 +252,240 @@ \subsection{BASIC 65 Variables}
-
-\screentextwide{C\% = C@\&(0,0)}
-
-T\% will then contain the display code of the character, while C\% has the colour index.
-
+A variable that stores a single value is also known as a {\em scalar
+variable}. Commodore BASIC also supports a type of variable that can store
+multiple values, called an {\em array}.
+
+The following example stores three string values in an array, then uses a {\bf
+FOR} loop to {\bf PRINT} a message for each element:
+
+\begin{screenoutput}
+10 DIM NA$(3)
+20 NA$(0) = "DEFT"
+30 NA$(1) = "GARDNERS"
+40 NA$(2) = "LYDON"
+50 FOR I=0 TO 3
+60 PRINT "HELLO, ";NA$(I);"!"
+70 NEXT I
+\end{screenoutput}
+
+Each value in an array is referenced by the name of the array variable and an
+integer index. For example, \screentext{AA(7)} refers to the element of the
+array named \screentext{AA()} with index 7. Indexes are ``zero-based:'' the
+first element in the array has an index of 0. The index can be a numeric
+expression, which can be a powerful way to operate on multiple elements of
+data.
+
+All values in an array must be of the same type. The type is indicated in the
+name of the variable, similar to scalar variables. \screentext{AA()} is an
+array of real numbers, \screentext{AA\$()} is an array of strings.
+
+Array variable names are considered separate from scalar variable names. The
+scalar variable \screentext{AA} has no relationship to the array variable
+\screentext{AA()}.
+
+Array variables must be declared before use with their maximum size. BASIC 65
+allocates the maximum amount of memory needed for the array. The {\bf DIM}
+command declares (``dimensions'') an array.
+
+An array can have multiple dimensions, each with its own index separated by a
+comma. The array must be declared with the maximum value for each dimension.
+Keep in mind that BASIC 65 allocates memory for the entire array, so large
+arrays may be constrained by available memory.
+
+\begin{screenoutput}
+DIM BO$(3,3)
+BO$(1,1) = "X"
+BO$(0,0) = "O"
+BO$(0,2) = "X"
+BO$(1,0) = "O"
+\end{screenoutput}
+
+\subsection{Screen Text and Colour Arrays}
+\label{sec:screentextandcolourarrays}
+\index{Screen Text and Colour Arrays}
+
+A BASIC 65 program can place text on the screen in several ways. The {\bf
+PRINT} command displays a string at the current cursor location, which is
+especially useful for terminal-like output. The {\bf CURSOR} command moves the
+cursor to a given position. A program can use these commands together to draw
+pictures or user interfaces.
+
+A program can access individual characters on the screen using the special
+built-in arrays \textbf{T@\&()} and \textbf{C@\&()}. These arrays are
+two-dimensional with indexes corresponding to the column and row of each
+character on the screen, starting from (0,0) at the top left corner.
+
+\textbf{T@\&(}column{\bf ,} row{\bf )} is the screen code of the character.
+Screen codes are not the same as PETSCII codes. Refer to a Commodore reference
+for a list of screen codes.
```

Comment:
ASIDE: Hehe yes, I look forward to the day we have screen codes in our mega65-book too, I'm tired of googling for them on the web :)

Feel like I recall you had a comment somewhere in here expressing similar sentiments :)

Reply:
yeah I even filed a bug so we don't forget :)